pyanaconda/controller.py

```diff
@@ -0,0 +1,124 @@
+        self._modules = set()
+        self._all_modules_added = False
+        self.init_done = Signal()
+        self._init_done_triggered = False
+        self._added_module_count = 0
+
+    @synchronized
+    def module_init_start(self, module):
+        """Tell the controller that a module has started initialization.
+
+        :param module: a module which has started initialization
+        """
+        if self._all_modules_added:
+            log.warning("Late module_init_start() from: %s", self)
+        elif module in self._modules:
+            log.warning("Module already marked as initializing: %s", module)
+        else:
+            self._added_module_count += 1
+            self._modules.add(module)
+
+    @synchronized
+    def all_modules_added(self):
+        """Tell the controller that all expected modules have started initialization.
+
+        Tell the controller that all expected modules have been registered
+        for initialization tracking (or have already been initialized)
+        and no more are expected to be added.
+
+        This is needed so that we don't prematurely trigger the init_done signal
+        when all known modules finish initialization while other modules have not
+        yet been added.
+        """
+        log.info("Initialization of all modules (%d) has been started.", self._added_module_count)
+        self._all_modules_added = True
+
+        # if all modules finished initialization before this was added then
+        # trigger the init_done signal at once
+        if not self._modules and not self._init_done_triggered:
+            self._trigger_init_done()
+
+    def module_init_done(self, module):
+        """Tell the controller that a module has finished initialization.
+
+        An if no more modules are being initialized trigger the init_done signal.
+
+        :param module: a module that has finished initialization
+        """
+        init_done = False
+        with self._lock:
+            # prevent the init_done signal from
+            # being triggered more than once
+            if self._init_done_triggered:
+                log.warning("Late module_init_done from module %s.", module)
+                return
+            if module in self._modules:
+                log.info("Module initialized: %s", module)
+                self._modules.discard(module)
+            else:
+                log.warning("Unknown module reported as initialized: %s", module)
+            # don't trigger the signal if all modules have not yet been added
+            if self._all_modules_added and not self._modules:
+                init_done = True
+                self._init_done_triggered = True
+
+        # we should be able to safely emit the signal out of the main lock
+        if init_done:
+            self._trigger_init_done()
+
+    def _trigger_init_done(self):
+        log.info("All modules have been initialized.")
+        self.init_done.emit()
+
+def get_controller(category_name):
```

Comment:
I don't know - that would basically mean storing mutable data in a class variable and that seems kinda weird. Also if I understand this correctly it would be called like this:

`controller.Controller.get_controller("foo")`

Which IMHO doesn't look any better than the current way.

Maybe we can rename the module to something else (`controllers`, etc.) or add a separate module (`spoke_startup`, `module_startup`, `module_init`, etc.) that has the dicts of  the `Controller` instances & the access functions while the `Controller` class still lives in the `controller` module.

Then the call could look like this:

controllers.get_controller("foo")
or
spoke_startup.get_controller("foo")
or
module_startup.get_controller("foo")
or 
module_init.get_controller("foo")

As for `controller.controllers.get()` I guess we could change the variable to something like `_controllers` (or even `__controllers`) if we want people to only use it via the access functions. and not directly.